### PR TITLE
Handle peer addition/removal in a right way

### LINKED ISSUE #160: Selected peer in the "Peers" tab gets unselected if any previous peer disconnects
On master(209974d379e5c2f32f9f691c992a2fb58be2abf2) the selected peer in the "Peers" tab gets unselected if any previous peer (in the upper rows) disconnects.

**Step 1** -- connected three peers, and "peerC" has been selected:
Screenshot from 2020-12-22 15-04-19

**Step 2** -- new connected peers ("peerD") do not hurt:
Screenshot from 2020-12-22 15-04-30

**Step 3** -- but disconnection of any peer from upper rows (e.g., "peerB") causes the loss of peer selection:
Screenshot from 2020-12-22 15-05-28

### LINKED ISSUE #191: User Agent header flickers when trying to resize any column
header_bug

System: Linux Mint 20.1 + Cinnamon DE + Qt 5.12.8

### PR DESCRIPTION
**Pull request by hebasto**
This PR makes `PeerTableModel` handle a peer addition/removal in a right way. See:
- https://doc.qt.io/qt-5/model-view-programming.html#inserting-and-removing-rows
- https://doc.qt.io/qt-5/model-view-programming.html#resizable-models

Fixes #160.

Fixes #191.

### CODE REVIEW

**Comment by hebasto** [comment]
Rebased 734dd5f95168886e12dc2f07dd91902b1242ffc3 -> d325799acdd8d010d5e4b2f77d9cc67d76d4ad28 ([pr164.01](https://github.com/hebasto/gui/commits/pr164.01) -> [pr164.02](https://github.com/hebasto/gui/commits/pr164.02)) due to the conflict with #162.

**Comment by hebasto** [comment]
Will be rebased after any progress in #161 :)

**Comment by hebasto** [comment]
Rebased d325799acdd8d010d5e4b2f77d9cc67d76d4ad28 -> f73d340abeedf7e583fe9b2a8b3685a745a9199e ([pr164.02](https://github.com/hebasto/gui/commits/pr164.02) -> [pr164.03](https://github.com/hebasto/gui/commits/pr164.03)) due to the conflict with #161.

**Comment by promag** [comment]
Why is this based on #18?

**Comment by hebasto** [comment]
> Why is this based on #18?

It fixes selection behavior and handling in general.

**Review by jarolrod — COMMENTED** [review]
tACK f73d340abeedf7e583fe9b2a8b3685a745a9199e

Confirming that the PR fixes the issue laid out in #160 on macOS 11.1 with Qt 5.15.2. 

On `master`: If I select a peer, and a peer above it is disconnected, then the peer I had selected becomes unselected. Below are screenshots exhibiting this behavior.

**Master:** I have selected `Peer ID 32`
peer-is-selected

**Master:** `Peer ID 32` is no longer selected because `Peer ID 29` is above it and disconnected
peer-becomes-deselected

on this `PR`: This behavior no longer occurs. If I select a peer, it remains selected even if a peer above it is disconnected.

**PR:** `Peer ID 14` is selected
pr-peer-selected

**PR:** `Peer ID 0` has disconnected. `Peer ID 14` remains selected.
pr-peer-remains-selected

**Comment by hebasto** [comment]
Rebased f73d340abeedf7e583fe9b2a8b3685a745a9199e -> 32ce58cb18acc6eb3ac3a2c650c38b7bfdb6ad8b ([pr164.03](https://github.com/hebasto/gui/commits/pr164.03) -> [pr164.04](https://github.com/hebasto/gui/commits/pr164.04)) due to the conflict with #180.

**Comment by hebasto** [comment]
Rebased 32ce58cb18acc6eb3ac3a2c650c38b7bfdb6ad8b -> a3d21d3dc3d9166a60f050352ea0eff2a8c2140b ([pr164.04](https://github.com/hebasto/gui/commits/pr164.04) -> [pr164.05](https://github.com/hebasto/gui/commits/pr164.05)) due to the conflict with #179.

**Review by jarolrod — COMMENTED** [review]
re-ACK a3d21d3dc3d9166a60f050352ea0eff2a8c2140b, tested on macOS 11.1 Qt 5.15.2 after rebase

**Comment by hebasto** [comment]
Rebased a3d21d3dc3d9166a60f050352ea0eff2a8c2140b -> fa95a2205e8c88b23e786a382e55f4534e7733ac ([pr164.05](https://github.com/hebasto/gui/commits/pr164.05) -> [pr164.06](https://github.com/hebasto/gui/commits/pr164.06)) due to the conflict with https://github.com/bitcoin/bitcoin/pull/21015.

**Comment by hebasto** [comment]
Rebased fa95a2205e8c88b23e786a382e55f4534e7733ac -> b7fe2b1691330166b478f0ced2dd1b6a8835729b ([pr164.06](https://github.com/hebasto/gui/commits/pr164.06) -> [pr164.07](https://github.com/hebasto/gui/commits/pr164.07)) due to the conflict with #166..

**Review by jarolrod — COMMENTED** [review]
re-ack b7fe2b1691330166b478f0ced2dd1b6a8835729b, tested on macOS 11.2 Qt 5.15.2 after rebase

File: src/qt/clientmodel.h [hunk 1]
```diff
@@ -54,6 +55,7 @@ class ClientModel : public QObject
     interfaces::Node& node() const { return m_node; }
     OptionsModel *getOptionsModel();
     PeerTableModel *getPeerTableModel();
+    PeerTableSortProxy* peerTableSortProxy();
```
Comment (Talkless) [inline src/qt/clientmodel.h:58]: nit: could be noexcept, since it just returns pointer. Though since other similar getters here does not use `noexcept`, adding that one kinda seems out of scope.

File: src/qt/peertablemodel.cpp [hunk 2]
```diff
@@ -223,35 +131,55 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
...
-    Q_EMIT layoutAboutToBeChanged();
-    priv->refreshPeers(m_node);
-    Q_EMIT layoutChanged();
-}
+    interfaces::Node::NodesStats nodes_stats;
+    m_node.getNodesStats(nodes_stats);
+    decltype(m_peers_data) new_peers_data;
+    new_peers_data.reserve(nodes_stats.size());
+    for (const auto& node_stats : nodes_stats) {
+        CNodeCombinedStats stats;
```
Comment (Talkless) [inline src/qt/peertablemodel.cpp:152]: These stats has quite a few members, and they all gonna be overwritten after construction. Since it's loop, consider initializing in this wat:
```
const CNodeCombinedStats stats{std::get<0>(node_stats), std::get<2>(node_stats), std::get<1>(node_stats)};
```

 It's most dissapointing that `QList` does not have `emplace()` here...

File: src/qt/peertablemodel.h [hunk 3]
```diff
@@ -74,22 +62,25 @@ class PeerTableModel : public QAbstractTableModel
...
 
 public Q_SLOTS:
     void refresh();
 
+Q_SIGNALS:
+    void changed();
+
 private:
+    //! Internal peer data structure.
+    QList<CNodeCombinedStats> m_peers_data{};
```
Comment (Talkless) [inline src/qt/peertablemodel.h:81]: nit: `{}` seems redundant, it's not fundamental type like int, pointer or bool to need this kind of explicit default initialization, constructor will be invoked regardless.

File: src/qt/peertablemodel.h [hunk 4]
```diff
@@ -74,22 +62,25 @@ class PeerTableModel : public QAbstractTableModel
 
     /** @name Methods overridden from QAbstractTableModel
         @{*/
-    int rowCount(const QModelIndex &parent) const override;
-    int columnCount(const QModelIndex &parent) const override;
-    QVariant data(const QModelIndex &index, int role) const override;
-    QVariant headerData(int section, Qt::Orientation orientation, int role) const override;
-    QModelIndex index(int row, int column, const QModelIndex &parent) const override;
+    int rowCount(const QModelIndex& parent = QModelIndex()) const override;
```
Comment (Talkless) [inline src/qt/peertablemodel.h:65]: nit: if one likes C++11 toys, one could write `= {}` instead of ` = QModelIndex()`.

File: src/qt/peertablesortproxy.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <qt/peertablemodel.h>
+#include <util/check.h>
+
+#include <QModelIndex>
+#include <QString>
+#include <QVariant>
+
+PeerTableSortProxy::PeerTableSortProxy(QObject* parent)
+    : QSortFilterProxyModel(parent)
```
Comment (Talkless) [inline src/qt/peertablesortproxy.cpp:15]: nit: since you use `{}` in `m_peer_table_sort_proxy{nullptr};`, maybe worth using `{}` in `QSortFilterProxyModel{parent}` too (meaning in the new code)?

File: src/qt/rpcconsole.cpp [hunk 6]
```diff
@@ -9,13 +9,14 @@
 #include <qt/rpcconsole.h>
 #include <qt/forms/ui_debugwindow.h>
 
+#include <chainparams.h>
+#include <interfaces/node.h>
```
Comment (Talkless) [inline src/qt/rpcconsole.cpp:13]: On My Machine™, it seems only `#include <interfaces/node.h>` is needed?

**Review by Talkless — CHANGES_REQUESTED** [review]
Concept ACK, right fix with nice cleanup.

File: src/qt/clientmodel.h (hunk 1 above)
Comment (hebasto) [reply src/qt/clientmodel.h:58]: Though since other similar getters here does not use `noexcept`, adding that one kinda seems out of scope.

Going to leave it as is.

File: src/qt/peertablemodel.h (hunk 3 above)
Comment (hebasto) [reply src/qt/peertablemodel.h:81]: Yes, it is redundant. Leaving as is to be explicit.

File: src/qt/peertablesortproxy.cpp (hunk 5 above)
Comment (hebasto) [reply src/qt/peertablesortproxy.cpp:15]: As stated in the commit description, the goal is to follow Qt [docs](https://doc.qt.io/qt-5/qabstracttablemodel.html#index).

File: src/qt/rpcconsole.cpp (hunk 6 above)
Comment (hebasto) [reply src/qt/rpcconsole.cpp:13]: That commit only adds `#include <qt/peertablesortproxy.h>`, and other `include`s are sorted only. I'd leave the need of the `interfaces/node.h` header out of this pr scope.

File: src/qt/peertablemodel.h (hunk 4 above)
Comment (hebasto) [reply src/qt/peertablemodel.h:65]: As stated in the commit description, the goal is to follow Qt [docs](https://doc.qt.io/qt-5/qabstracttablemodel.html#index).

File: src/qt/peertablesortproxy.cpp (hunk 5 above)
Comment (hebasto) [reply src/qt/peertablesortproxy.cpp:15]: This commit belongs to the #18 on which this one is based.

**Comment by hebasto** [comment]
Updated b7fe2b1691330166b478f0ced2dd1b6a8835729b -> 7bb0dc97ab08ebcba84b54031874c0648367d756 ([pr164.07](https://github.com/hebasto/gui/commits/pr164.07) -> [pr164.08](https://github.com/hebasto/gui/commits/pr164.08), [diff](https://github.com/hebasto/gui/compare/pr164.07..pr164.08)):

- addressed @Talkless' [comment](https://github.com/bitcoin-core/gui/pull/164#discussion_r592667897)

File: src/qt/peertablemodel.cpp (hunk 2 above)
Comment (hebasto) [reply src/qt/peertablemodel.cpp]: Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-797463817).

File: src/qt/rpcconsole.cpp (hunk 6 above)
Comment (Talkless) [reply src/qt/rpcconsole.cpp:13]: Sorry, missed the fact that it's a sort.

File: src/qt/peertablemodel.cpp [hunk 7]
```diff
@@ -223,35 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
...
 {
-    Q_EMIT layoutAboutToBeChanged();
-    priv->refreshPeers(m_node);
-    Q_EMIT layoutChanged();
-}
+    interfaces::Node::NodesStats nodes_stats;
+    m_node.getNodesStats(nodes_stats);
+    decltype(m_peers_data) new_peers_data;
+    new_peers_data.reserve(nodes_stats.size());
+    for (const auto& node_stats : nodes_stats) {
```
Comment (Talkless) [inline src/qt/peertablemodel.cpp:151]: Ugh, my suggestion was not good enought, you can actually move values out of that local `nodes_stats`:
```
    for (auto&& node_stats : nodes_stats) {
        const CNodeCombinedStats stats{std::get<0>(std::move(node_stats)), 
                                       std::get<2>(std::move(node_stats)),
                                       std::get<1>(std::move(node_stats))};
```
(note loop variable is no longer cost reference!)

Or even with an algorithm (also moves, at least it does on local test), if one would care about "No raw loops" mantra :) 
```
    std::transform(nodes_stats.begin(), nodes_stats.end(), std::back_inserter(new_peers_data), [](auto&& i) {
        return CNodeCombinedStats{std::get<0>(std::move(i)), std::get<2>(std::move(i)), std::get<1>(std::move(i))};
    });

```
Though moving `std::transform` seems fishy, not sure if it's "legal" to move source... But gcc 10.2.1 does that.

Comment (hebasto) [reply src/qt/peertablemodel.cpp:151]: > Ugh, my suggestion was not good enought...

It seems like a kind of premature optimization, no?

Comment (Talkless) [reply src/qt/peertablemodel.cpp:151]: Maybe, probably. Since there will be mostly on up to ~128 nodes by default, "nobody" would care if there are some extra copies I guess, and `std::transform` variant do look particularly "noisy" (i.e. "optimization" not worth the effort/readability cost).

File: src/qt/clientmodel.h (hunk 1 above)
Comment (promag) [reply src/qt/clientmodel.h:58]: FWIW I've suggested that this should not be "global" but rather instanced for each view (currently one in console dialog).

**Comment by hebasto** [comment]
Rebased 7bb0dc97ab08ebcba84b54031874c0648367d756 -> ec3075d7342b65bdd72e34f5705f6cdc8a948b67 ([pr164.08](https://github.com/hebasto/gui/commits/pr164.08) -> [pr164.09](https://github.com/hebasto/gui/commits/pr164.09)) due to the merging of #18.

**Review by Talkless — APPROVED** [review]
tACK ec3075d7342b65bdd72e34f5705f6cdc8a948b67, tested on Debian Sid with Qt 5.15.2. 

Tested using two bitcoind's connected to PR's bitcoin-qt in regtest mode. Did some reconnects, and so far table seems to work fine.

**Comment by hebasto** [comment]
Rebased ec3075d7342b65bdd72e34f5705f6cdc8a948b67 -> e102c90e41fb82e534c155bd7cfb9d0e9852c092 ([pr164.09](https://github.com/hebasto/gui/commits/pr164.09) -> [pr164.10](https://github.com/hebasto/gui/commits/pr164.10)) due to the conflict with #311.

**Review by promag — COMMENTED** [review]
Tested ACK e102c90e41fb82e534c155bd7cfb9d0e9852c092 on macos.

**Comment by jonatack** [comment]
Concept ACK. If this fixes what I think it does, it would be good to see it in v22 as a bug fix.

File: src/qt/peertablemodel.cpp [hunk 8]
```diff
@@ -172,19 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
...
+        const CNodeCombinedStats stats{std::get<0>(node_stats), std::get<2>(node_stats), std::get<1>(node_stats)};
+        new_peers_data.append(stats);
+    }
+
+    // Handle a new peer addition or an old peer in a right way. See:
```
Comment (jarolrod) [inline src/qt/peertablemodel.cpp:156]: nit, or something similar
```suggestion
    // Handle new peer addition or removal as suggested in Qt Docs. See:
```

File: src/qt/peertablemodel.cpp [hunk 9]
```diff
@@ -172,19 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
...
+    // - https://doc.qt.io/qt-5/model-view-programming.html#inserting-and-removing-rows
+    // - https://doc.qt.io/qt-5/model-view-programming.html#resizable-models
+    // We use the fact that interfaces::Node::getNodesStats
```
Comment (jarolrod) [inline src/qt/peertablemodel.cpp:159]:
```suggestion
    // We take advantage of the fact that the std::vector returned by  interfaces::Node::getNodesStats
```

File: src/qt/peertablemodel.cpp [hunk 10]
```diff
@@ -172,19 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
...
+    // returns std::vector naturally sorted by nodeid.
```
Comment (jarolrod) [inline src/qt/peertablemodel.cpp:160]:
```suggestion
    // is sorted by nodeid.
```

**Review by jarolrod — COMMENTED** [review]
ACK e102c90e41fb82e534c155bd7cfb9d0e9852c092

Tested functionality on `Arch Linux Qt 5.15.2`, `macOS 11.3 Qt 5.15.2`, and cross-compile from Linux to `Windows 10`.

## Notes on Commits:
- [x] efb7e5aa962d4a4047061996bbb50b6da4592cbc
  - code review ACK cross-checked with Qt docs for [QAbstractTableModel](https://doc.qt.io/qt-5/qabstracttablemodel.html) and [QAbstractItemModel](https://doc.qt.io/qt-5/qabstractitemmodel.html)
- [x] 1b66f6e556631a1a2d89aefba70a79894bd14fcd
  - code review ACK, the `PeerTablePriv` class would no longer be needed.
- [x] e102c90e41fb82e534c155bd7cfb9d0e9852c092
  - code review ACK, patch looks correct in its implementation in following recommendations from Qt docs
  - some nits () 
## Screenshots:

The below screenshots show that this PR will close the mutated version of #160 (see, https://github.com/bitcoin-core/gui/issues/160#issuecomment-855515389). While I am not including videos showing that this PR also fixes #191, I can confirm that it does on all three tested platforms.

### Arch Linux

#### master

| Select Peer  6   | Disconnect Peer 0, Peer 7 is now selected ❌ |
| --------------- | --------------------------------------------- |
| master0start | master-disconnect |

#### PR

| Select Peer  9   | Disconnect Peer 1, Peer 9 remains selected ✅  |
| --------------- | ----------------------------------------------- |
| linux-pr-start | linux-pr-disconnect |

### macOS 11.3

#### master

| Select Peer  6   | Disconnect Peer 0, Peer 7 is now selected ❌ |
| --------------- | --------------------------------------------- |
| Screen Shot 2021-06-06 at 9 17 11 PM | Screen Shot 2021-06-06 at 9 17 48 PM |

#### PR
| Select Peer  8   | Disconnect Peer 0, Peer 8 remains selected ✅ |
| --------------- | ---------------------------------------------- |
| macOS-pr-start | macOS-pr-disconnect |

### Windows 10

Note: Difference between UI on master and PR is because master includes ab86ac7739b27f2c45ae72ce4c4ebecf324b3e90

#### master

| Select Peer  4   | Disconnect Peer 0, Peer 6 is now selected ❌ |
| --------------- | --------------------------------------------- |
| windows-master-start | windows-master-disconnect |

#### PR

| Select Peer  5  | Disconnect Peer 1, Peer 5 remains selected ✅ |
| --------------- | ---------------------------------------------- |
| windows-start-pr | windows-disconnect-pr |

**Comment by hebasto** [comment]
Updated e102c90e41fb82e534c155bd7cfb9d0e9852c092 -> ecbd91153875c8cdd5b92b840afc116f65e457fb ([pr164.10](https://github.com/hebasto/gui/commits/pr164.10) -> [pr164.11](https://github.com/hebasto/gui/commits/pr164.11), [diff](https://github.com/hebasto/gui/compare/pr164.10..pr164.11)):

- addressed recent @jarolrod's comments

File: src/qt/peertablemodel.cpp (hunk 8 above)
Comment (hebasto) [reply src/qt/peertablemodel.cpp]: Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-855996526).

File: src/qt/peertablemodel.cpp (hunk 9 above)
Comment (hebasto) [reply src/qt/peertablemodel.cpp]: [Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-855996526).

File: src/qt/peertablemodel.cpp (hunk 10 above)
Comment (hebasto) [reply src/qt/peertablemodel.cpp]: [Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-855996526).

**Review by promag — COMMENTED** [review]
reACK ecbd91153875c8cdd5b92b840afc116f65e457fb just improvements to the comment since last review.

**Comment by jarolrod** [comment]
re-ACK ecbd911

Only changes since last review are implementation of comment nits
